docs/proposals/20210210-insulate-users-from-kubeadm-API-changes.md

```diff
@@ -0,0 +1,226 @@
+
+### Requirements
+
+R1 - avoid breaking changes in v1alpha3
+R2 - ensure a clean v1alpha3 to v1alpha4 upgrade
+
+### Implementation Details/Notes/Constraints
+
+#### Background info about kubeadm API version
+
+kubeadm v1beta1 types:
+
+- Introduced in v1.13.
+- Deprecation cycle started with v1.17.
+- Removal scheduled for v1.20 but then postponed to v1.21.
+
+kubeadm v1beta2 types:
+
+- Introduced in v1.15.
+- Changes from the previous version are minimal:
+  - Support for IgnorePreflightErrors into the NodeRegistrationOptions.
+  - Support for CertificateKey into InitConfiguration and JoinConfiguration
+    configuration (not relevant for Cluster API because it is not using the
+    automatic certificate copy feature).
+  - improved serialization support (in practice a set of `omitempty` fixes).
+
+#### Background info about kubeadm types into the KubeadmConfig/KubeadmControlPlane specs
+
+Given the fact that importing kubead (which is part of the Kubernetes codebase)
+in Cluster API is impratical, Cluster API hosts a mirror of kubeadm API types.
+
+kubeadm v1beta1 mirror-types:
+
+- Hosted in `bootstrap/kubeadm/types/v1beta1`.
+- Diverged from the original v1beta1 types for better CRD support (in practice
+  a set of `+optional` fixes, few `omitempty` differences).
+
+kubeadm v1beta2 mirror-types:
+
+- Hosted in `bootstrap/kubeadm/types/v1beta2`.
+- Currently, not used in the Cluster API codebase.
+- Does not include changes for better CRD support introduced in kubeadm v1beta1
+  mirror-types.
+
+#### Cluster API v1alpha3 changes
+
+Changes to cluster API v1alpha3 release should be minimal and no breaking change
+should be introduced while implemeting this proposal.
+
+__Alternative 1:__
+
+Keep kubeadm v1beta1 types as a Hub type; implement conversion to kubeadm API
+version f(Kubernetes Version) when generating the kubeadm config for init/join.
+
+This alternative is the more clean, robust and forward looking, but it requires
+much more work than alternative 2.
+
+__Alternative 2:__
+
+Keep kubeadm v1beta1 types as a Hub type; only change the apiVersion to
+`kubeadm.k8s.io/v1beta2` in the generated kubeadm config for init/join.
+
+This alternative is based on following considerations:
+
+- kubeadm v1beta1 mirror types are "compatible" with kubeadm v1beta2 types
+  - Having support for IgnorePreflightErrors into the NodeRegistrationOptions
+    is not required for Cluster API v1alpha3
+  - The automatic certificate copy feature is not used by Cluster API
+  - Improved serialization support has been already applied to the kubeadm
+    v1beta1 mirror-types (and further extended).
+- the minimal Kubernetes version supported by Cluster API is v1.16, and kubeadm
+  v1.16 could work with v1beta2 types. Same for all the other versions up to latest(1.20)
+  and next (1.21).
+- limitations: this approach is not future proof, and it should be reconsidered
+  whenever a new version of kubeadm types is created while v1alpha3 is still supported.
```

Comment:
it does feel like a temporary hack.

Reply:
there are pro and cons.
it has limitations and it is not future proof but it limits the impact/the risks on the v1alpha3 branch.
but let's see other opinions on the two alternatives